framework/src/utils/MooseCoordTransform.C

```diff
@@ -0,0 +1,276 @@
+    mooseError("Unsupported manual angle prescription in 'MooseCoordTransform::setRotation'");
+}
+
+void
+MooseCoordTransform::setLengthUnitsPerMeter(const Real length_units_per_meter)
+{
+  const auto scale = 1 / length_units_per_meter;
+
+  _scale =
+      std::make_unique<RealTensorValue>(RealTensorValue(scale, 0, 0, 0, scale, 0, 0, 0, scale));
+}
+
+void
+MooseCoordTransform::setCoordinateSystem(const Moose::CoordinateSystemType coord_type)
+{
+  _coord_type = coord_type;
+}
+
+MooseCoordTransform::MooseCoordTransform(const InputParameters & params)
+{
+  //
+  // coordinate system transformation. If we have multiple different coordinate system types in our
+  // problem, we take note of it because that can cause issues if there is a non-Cartesian
+  // destination coordinate system
+  //
+  const auto & mesh = *params.getCheckedPointerParam<MooseMesh *>("mesh");
+  const auto & coord_sys = mesh.getCoordSystem();
+  std::unordered_set<Moose::CoordinateSystemType> coord_types;
+  auto map_it = coord_sys.begin();
+  // It's possible that the mesh is not in a complete state
+  if (map_it == coord_sys.end())
+    setCoordinateSystem(Moose::COORD_XYZ);
+  else
+    setCoordinateSystem(map_it->second);
+  for (; map_it != coord_sys.end(); ++map_it)
+    coord_types.insert(map_it->second);
+
+  _has_different_coord_sys = coord_types.size() > 1;
+
+  //
+  // rotation
+  //
+  const bool has_alpha = params.isParamValid("alpha_rotation");
+  const bool has_beta = params.isParamValid("beta_rotation");
+  const bool has_gamma = params.isParamValid("gamma_rotation");
+  const auto & up_direction = params.get<MooseEnum>("what_is_up");
+
+  const Direction rz_z_axis =
+      (_coord_type == Moose::COORD_RZ)
+          ? Direction(static_cast<unsigned int>(int(params.get<MooseEnum>("rz_coord_axis"))))
+          : Y;
+
+  if (has_alpha || has_beta || has_gamma)
+  {
+    if (up_direction.isValid())
+      mooseError("Cannot simultaneously set rotation angles as well as an up direction");
+
+    const auto alpha = (has_alpha ? params.get<Real>("alpha_rotation") : Real(0));
+    const auto beta = (has_beta ? params.get<Real>("beta_rotation") : Real(0));
+    const auto gamma = (has_gamma ? params.get<Real>("gamma_rotation") : Real(0));
+
+    if (_coord_type == Moose::COORD_RZ)
+      setRotation(alpha, beta, gamma, rz_z_axis);
+    else
+      setRotation(alpha, beta, gamma);
+  }
+  else if (up_direction.isValid())
+    setRotation(Direction(static_cast<unsigned int>(int(up_direction))), rz_z_axis);
+
+  //
+  // Scale
+  //
+  if (params.isParamValid("length_units_per_meter"))
+    setLengthUnitsPerMeter(params.get<Real>("length_units_per_meter"));
+}
+
+Point
+MooseCoordTransform::operator()(const Point & point) const
+{
+  Point ret(point);
+  if (_rotate)
+    ret = (*_rotate) * ret;
+  if (_scale)
+    ret = (*_scale) * ret;
+
+  // If this shows up in profiling we can make _translation a pointer
+  ret += _translation;
```

Comment:
since this does not commute with the others we ll have to be clear that it's performed after